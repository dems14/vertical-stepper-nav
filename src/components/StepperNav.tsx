import * as React from 'react';
import {Step} from './Step';
import {Separator} from './Separator';

export interface IStepDescription {
  stepContent: () => JSX.Element;
  stepStateColor?: string;
  stepStatusCircleSize?: number;
  onClickHandler?: () => void | undefined;
}

export interface IStepperNavProps {
  steps: IStepDescription[];
}

export const StepperNav = (props: IStepperNavProps): JSX.Element => {
  return (
    <nav>
      {props.steps.map(
        (
          {stepContent, stepStateColor, onClickHandler, stepStatusCircleSize},
          index,
        ) => (
          <div key={index}>
            <Step
              stepContent={stepContent}
              statusColor={stepStateColor}
              onClickHandler={onClickHandler}
              statusCircleSize={stepStatusCircleSize}
            />
            {index !== props.steps.length - 1 && (
              <div
                style={{
                  paddingLeft: `${(stepStatusCircleSize ?? 16) / 2 + 1}px`,
                }}>
                <Separator />
              </div>
            )}
          </div>
        ),
      )}
    </nav>
  );
};
