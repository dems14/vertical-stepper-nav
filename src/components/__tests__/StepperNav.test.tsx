/* eslint-disable no-undef */
import * as React from 'react';
import '@testing-library/jest-dom';
import {StepperNav} from '../StepperNav';
import {render, fireEvent, screen} from '@testing-library/react';

describe('StepperNav tests', () => {
  test('basic snapshot', () => {
    const {asFragment} = render(
      <StepperNav
        steps={[
          {
            stepContent: () => <div style={{fontSize: 14}}>Step 1</div>,
            stepStatusCircleSize: 16,
            stepStateColor: 'darkblue',
          },
          {
            stepContent: () => <div style={{fontSize: 14}}>Step 2</div>,
            stepStatusCircleSize: 16,
          },
          {
            stepContent: () => <div style={{fontSize: 14}}>Step 3</div>,
            stepStateColor: 'green',
          },
          {
            stepContent: () => <div style={{fontSize: 14}}>Step 3</div>,
          },
        ]}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('user can invoke the onClick handler by clicking on a step', () => {
    const onClickHandlerMock = jest.fn();
    render(
      <StepperNav
        steps={[
          {
            stepContent: () => <div style={{fontSize: 14}}>Step 1</div>,
            stepStatusCircleSize: 16,
            stepStateColor: 'darkblue',
            onClickHandler: onClickHandlerMock,
          },
          {
            stepContent: () => <div style={{fontSize: 14}}>Step 2</div>,
            stepStatusCircleSize: 16,
            stepStateColor: 'darkgreen',
          },
          {
            stepContent: () => <div style={{fontSize: 14}}>Step 3</div>,
            stepStatusCircleSize: 16,
            stepStateColor: 'red',
          },
        ]}
      />,
    );
    fireEvent.click(screen.getByText(/Step 1/i));
    expect(onClickHandlerMock).toHaveBeenCalledTimes(1);
  });
});
