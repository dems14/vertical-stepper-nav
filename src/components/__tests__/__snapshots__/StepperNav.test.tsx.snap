// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepperNav tests basic snapshot 1`] = `
<DocumentFragment>
  <nav>
    <div>
      <div
        role="button"
        style="display: inline-flex; flex-wrap: wrap; gap: 12px; padding: 2px; cursor: pointer;"
        tabindex="0"
      >
        <div>
          <div
            style="border-radius: 16px; width: 16px; height: 16px; border: 2px solid #e1e1e1; background: darkblue;"
          />
        </div>
        <div
          style="padding-bottom: 2px;"
        >
          <div
            style="font-size: 14px;"
          >
            Step 1
          </div>
        </div>
      </div>
      <div
        style="padding-left: 9px;"
      >
        <div
          style="height: 5vh; width: 2px; border: 1px solid #e1e1e1; background: rgb(225, 225, 225);"
        />
      </div>
    </div>
    <div>
      <div
        role="button"
        style="display: inline-flex; flex-wrap: wrap; gap: 12px; padding: 2px; cursor: pointer;"
        tabindex="0"
      >
        <div>
          <div
            style="border-radius: 16px; width: 16px; height: 16px; border: 2px solid #e1e1e1; background: white;"
          />
        </div>
        <div
          style="padding-bottom: 2px;"
        >
          <div
            style="font-size: 14px;"
          >
            Step 2
          </div>
        </div>
      </div>
      <div
        style="padding-left: 9px;"
      >
        <div
          style="height: 5vh; width: 2px; border: 1px solid #e1e1e1; background: rgb(225, 225, 225);"
        />
      </div>
    </div>
    <div>
      <div
        role="button"
        style="display: inline-flex; flex-wrap: wrap; gap: 12px; padding: 2px; cursor: pointer;"
        tabindex="0"
      >
        <div>
          <div
            style="border-radius: 16px; width: 16px; height: 16px; border: 2px solid #e1e1e1; background: green;"
          />
        </div>
        <div
          style="padding-bottom: 2px;"
        >
          <div
            style="font-size: 14px;"
          >
            Step 3
          </div>
        </div>
      </div>
      <div
        style="padding-left: 9px;"
      >
        <div
          style="height: 5vh; width: 2px; border: 1px solid #e1e1e1; background: rgb(225, 225, 225);"
        />
      </div>
    </div>
    <div>
      <div
        role="button"
        style="display: inline-flex; flex-wrap: wrap; gap: 12px; padding: 2px; cursor: pointer;"
        tabindex="0"
      >
        <div>
          <div
            style="border-radius: 16px; width: 16px; height: 16px; border: 2px solid #e1e1e1; background: white;"
          />
        </div>
        <div
          style="padding-bottom: 2px;"
        >
          <div
            style="font-size: 14px;"
          >
            Step 3
          </div>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
